ctxlogger/CtxLogger.cs

```diff
@@ -19,21 +28,161 @@ public CtxLoggerInterceptor(Serilog.ILogger logger)
+                {
+                    fieldMap.Remove(fieldName);
+                    continue;
+                } 
+            }
+
+            if (fieldMap[fieldName] is Dictionary<string, object> subMap && 
+                fieldDescriptor.FieldType == FieldType.Message)
+            {
+                var subMessageDescriptor = fieldDescriptor.MessageType;
+                if (subMessageDescriptor != null)
+                {
+                    fieldMap[fieldName] = FilterLoggableFields(subMap, subMessageDescriptor);
+                }
+            }
+        }
+        return fieldMap;
+    }
 }
+
+public static class JsonHelper
+{
+    public static Dictionary<string, object> ConvertJObjectToDictionary(JObject jObject)
+    {
+        var result = new Dictionary<string, object>();
+        foreach (var property in jObject.Properties())
+        {
+            if (property.Value is JObject nestedJObject)
+            {
+                result[property.Name] = ConvertJObjectToDictionary(nestedJObject);
+            }
+            else if (property.Value is JArray nestedArray)
+            {
+                result[property.Name] = ConvertJArrayToList(nestedArray);
+            }
+            else
+            {
+                result[property.Name] = property.Value?.ToObject<object>();
+            }
+        }
+        return result;
+    }
+
+    public static List<object> ConvertJArrayToList(JArray jArray)
+    {
+        var result = new List<object>();
+        foreach (var item in jArray)
+        {
+            if (item != null)
+            {
+                result.Add(item.ToObject<object>()!);
+            }
+        }
+        return result;
+    }
+}
+
+public static class LoggerExtensions
+{
+    public static ILogger WithCtx(this ILogger logger, ServerCallContext context,
+        [CallerFilePath] string callerFilePath = "",
+        [CallerLineNumber] int callerLineNumber = 0,
+        [CallerMemberName] string callerMemberName = "")
+    {
+        // Extract JSON string from request headers
+        var json = context.RequestHeaders.GetValue("ctxlog-data");
+
+        // Check if json is null or empty
+        if (string.IsNullOrEmpty(json))
+        {
+            return logger;
+        }
+
+        // Deserialize the JSON string back to a dictionary with type information
+        var dictionary = JsonConvert.DeserializeObject<Dictionary<string, object>>(json, new JsonSerializerSettings
+        {
+            TypeNameHandling = TypeNameHandling.Auto
+        });
+
+        // Add dictionary properties to the log context
+        if (dictionary != null)
+        {
+            foreach (var kvp in dictionary)
+            {
+                logger = logger.ForContext(kvp.Key, kvp.Value, destructureObjects: true);
+            }
+        }
+
+        // Add caller information to the log context
+        var location = new
+        {
+            function = callerMemberName,
```

Comment:
How are we able to access this filename, line number, etc of where the log came from in the middleware? Add any useful links you may have used for this

Reply:
added a link about Caller Information Attributes